### DISCUSSION TITLE
Duplicate port warning in controller deployment

### DISCUSSION THREAD

**Discussion by gmarcy** [topolvm/topolvm#801] [Q&A] [answered]
I am seeing this message when I restart the controller deployment
```
Warning: spec.template.spec.containers[4].ports[0]: duplicate port definition with spec.template.spec.containers[0].ports[1]
```

This appears to have been introduced when PR [709](https://github.com/topolvm/topolvm/pull/709) was merged.

The deployment has two references to port 9808, one [here](https://github.com/topolvm/topolvm/blob/main/charts/topolvm/templates/controller/deployment.yaml#L80) and the second [here](https://github.com/topolvm/topolvm/blob/main/charts/topolvm/templates/controller/deployment.yaml#L240).

Is this intentional or should the first definition of the port under the controller container be removed now that the port has been added explicitly to the livenessprobe container definition?

**Comment by toshipp** [comment, marked as answer]
Thank you for reporting the issue.
It is NOT intentional. 
the short answer: we should retain the first and remove the second.
the long answer:
topolvm-controller is a CSI controller plugin that communicates with kubelet or others via the unix domain socket. Since kubernetes cannot use unix domain sockets for livenessprobe, the livenessprobe sidecar is introduced. We want to check the controller liveness, so the port for livenessprobe is defined in the controller itself, but the sidecar provides the actual port. This is the reason we should do it.

If you send a PR, we are happy to review it.

**Reply by gmarcy** [reply]
@toshipp I started looking at the code changes to produce a PR and I don't believe that the 9808 port is known to the code running in the controller container at all.  I can find the webhookAddr (9443), metricsAddr (8080) and healthAddr (8081) explicitly mentioned in the topolvm source.  The only reference I can find to port 9808 is in the source for the livenessprobe container image, for example [here](https://github.com/kubernetes-csi/livenessprobe/blob/master/cmd/livenessprobe/main.go#L41).  Since that container is running in the controller pod it seems more appropriate for the definition for port 9808 to come from the livenessprobe container in the Deployment manifest.  Any objections?

**Reply by toshipp** [reply]
@gmarcy 
We refer to [the official livenessprobe document](https://github.com/kubernetes-csi/livenessprobe/blob/master/README.md). It defines the port in the driver container.

**Reply by gmarcy** [reply]
interesting.  will follow that guidance then.  thanks!

**Reply by gmarcy** [reply]
opened a PR here - https://github.com/topolvm/topolvm/pull/812
